File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/MailboxTest.java
/****************************************************************
 * Licensed to the Apache Software Foundation (ASF) under one   *
 * or more contributor license agreements.  See the NOTICE file *
 * distributed with this work for additional information        *
 * regarding copyright ownership.  The ASF licenses this file   *
 * to you under the Apache License, Version 2.0 (the            *
 * "License"); you may not use this file except in compliance   *
 * with the License.  You may obtain a copy of the License at   *
 *                                                              *
 *   http://www.apache.org/licenses/LICENSE-2.0                 *
 *                                                              *
 * Unless required by applicable law or agreed to in writing,   *
 * software distributed under the License is distributed on an  *
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
 * KIND, either express or implied.  See the License for the    *
 * specific language governing permissions and limitations      *
 * under the License.                                           *
 ****************************************************************/

package org.apache.james.jmap.model.mailbox;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.Optional;

import org.apache.james.mailbox.inmemory.InMemoryId;
import org.junit.Test;

public class MailboxTest {

    @Test(expected=NullPointerException.class)
    public void idShouldThrowWhenIdIsNull() {
        Mailbox.builder()
            .id(null);
    }

    @Test(expected=NullPointerException.class)
    public void nameShouldThrowWhenNameIsNull() {
        Mailbox.builder()
            .name(null);
    }

    @Test(expected=IllegalStateException.class)
    public void buildShouldThrowWhenIdIsNull() {
        Mailbox.builder().build();
    }

    @Test(expected=IllegalStateException.class)
    public void buildShouldThrowWhenNameIsNull() {
        Mailbox.builder()
            .id(InMemoryId.of(1))
            .build();
    }

    @Test(expected=IllegalStateException.class)
    public void buildShouldThrowWhenNameIsEmpty() {
        Mailbox.builder()
            .id(InMemoryId.of(1))
            .name("")
            .build();
    }

    @Test
    public void buildShouldWork() {
        Mailbox expectedMailbox = new Mailbox(InMemoryId.of(1), "name", Optional.of(InMemoryId.of(0)), Optional.of(Role.DRAFTS), SortOrder.of(123),
                true, true, true, true, true, true, true, 456, 789, 741, 852, Rights.EMPTY);

        Mailbox mailbox = Mailbox.builder()
            .id(InMemoryId.of(1))
            .name("name")
            .parentId(InMemoryId.of(0))
            .role(Optional.of(Role.DRAFTS))
            .sortOrder(SortOrder.of(123))
            .mustBeOnlyMailbox(true)
            .mayReadItems(true)
            .mayAddItems(true)
            .mayRemoveItems(true)
            .mayCreateChild(true)
            .mayRename(true)
            .mayDelete(true)
            .totalMessages(456)
            .unreadMessages(789)
            .totalThreads(741)
            .unreadThreads(852)
            .build();

        assertThat(mailbox).isEqualToComparingFieldByField(expectedMailbox);
    }

    @Test
    public void parentIdDefaultValueIsEmpty() {
        Mailbox mailbox = Mailbox.builder()
            .id(InMemoryId.of(0))
            .name("name")
            .build();

        assertThat(mailbox.getParentId()).isEmpty();
    }
}

File: server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/mail/MailboxTest.scala
/****************************************************************
 * Licensed to the Apache Software Foundation (ASF) under one   *
 * or more contributor license agreements.  See the NOTICE file *
 * distributed with this work for additional information        *
 * regarding copyright ownership.  The ASF licenses this file   *
 * to you under the Apache License, Version 2.0 (the            *
 * "License"); you May not use this file except in compliance   *
 * with the License.  You May obtain a copy of the License at   *
 *                                                              *
 * http://www.apache.org/licenses/LICENSE-2.0                   *
 *                                                              *
 * Unless required by applicable law or agreed to in writing,   *
 * software distributed under the License is distributed on an  *
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
 * KIND, either express or implied.  See the License for the    *
 * specific language governing permissions and limitations      *
 * under the License.                                           *
 * ***************************************************************/

package org.apache.james.jmap.mail

import eu.timepit.refined.auto._
import org.apache.james.core.Username
import org.apache.james.mailbox.Role
import org.apache.james.mailbox.model.TestId
import org.scalatest.matchers.must.Matchers
import org.scalatest.wordspec.AnyWordSpec

import scala.compat.java8.OptionConverters._

class MailboxTest extends AnyWordSpec with Matchers {
  "sortOrder" should  {
    "be comparable" in {
      SortOrder.apply(4L).compare(SortOrder.apply(3L)) must equal(1L)
      SortOrder.apply(4L).compare(SortOrder.apply(4L)) must equal(0L)
      SortOrder.apply(4L).compare(SortOrder.apply(5L)) must equal(-1L)
    }
  }

  "namespace" should  {
    "return None when personal" in {
      MailboxNamespace.personal.owner.isEmpty must be(true)
    }
    "return owner when delegated" in {
      val username = Username.of("bob")
      MailboxNamespace.delegated(username).owner must be(Some(username))
    }
  }

  "mailbox hasRole" should  {
    "return false when None " in {
      Mailbox(
        id = TestId.of(42L),
        name = "Inbox",
        parentId = None,
        role = None,
        SortOrder.apply(3L),
        TotalEmails(3L),
        UnreadEmails(4L),
        TotalThreads(5L),
        UnreadThreads(6L),
        MailboxRights(MayReadItems(true),
          MayAddItems(true),
          MayRemoveItems(true),
          MaySetSeen(true),
          MaySetKeywords(true),
          MayCreateChild(true),
          MayRename(true),
          MayDelete(true),
          MaySubmit(true)),
        IsSubscribed(true),
        MailboxNamespace.personal,
        Rights.EMPTY,
        Quotas(Map()))
        .hasRole(Role.INBOX) must be(false)
    }
    "return false when different" in {
      Mailbox(
        id = TestId.of(42L),
        name = "Inbox",
        parentId = None,
        role = Some(Role.OUTBOX),
        SortOrder.apply(3L),
        TotalEmails(3L),
        UnreadEmails(4L),
        TotalThreads(5L),
        UnreadThreads(6L),
        MailboxRights(MayReadItems(true),
          MayAddItems(true),
          MayRemoveItems(true),
          MaySetSeen(true),
          MaySetKeywords(true),
          MayCreateChild(true),
          MayRename(true),
          MayDelete(true),
          MaySubmit(true)),
        IsSubscribed(true),
        MailboxNamespace.personal,
        Rights.EMPTY,
        Quotas(Map()))
        .hasRole(Role.INBOX) must be(false)
    }
    "return true when equals" in {
      Mailbox(
        id = TestId.of(42L),
        name = "Inbox",
        parentId = None,
        role = Some(Role.INBOX),
        SortOrder.apply(3L),
        TotalEmails(3L),
        UnreadEmails(4L),
        TotalThreads(5L),
        UnreadThreads(6L),
        MailboxRights(MayReadItems(true),
          MayAddItems(true),
          MayRemoveItems(true),
          MaySetSeen(true),
          MaySetKeywords(true),
          MayCreateChild(true),
          MayRename(true),
          MayDelete(true),
          MaySubmit(true)),
        IsSubscribed(true),
        MailboxNamespace.personal,
        Rights.EMPTY,
        Quotas(Map())).hasRole(Role.INBOX) must be(true)
    }
  }

  "mailbox hasSystemRole" should  {
    "return false when None" in {
      Mailbox(
        id = TestId.of(42L),
        name = "Inbox",
        parentId = None,
        role = None,
        SortOrder.apply(3L),
        TotalEmails(3L),
        UnreadEmails(4L),
        TotalThreads(5L),
        UnreadThreads(6L),
        MailboxRights(MayReadItems(true),
          MayAddItems(true),
          MayRemoveItems(true),
          MaySetSeen(true),
          MaySetKeywords(true),
          MayCreateChild(true),
          MayRename(true),
          MayDelete(true),
          MaySubmit(true)),
        IsSubscribed(true),
        MailboxNamespace.personal,
        Rights.EMPTY,
        Quotas(Map()))
        .hasSystemRole must be(false)
    }
    "return false when not system" in {
      Mailbox(
        id = TestId.of(42L),
        name = "Inbox",
        parentId = None,
        Role.from("any").asScala,
        SortOrder.apply(3L),
        TotalEmails(3L),
        UnreadEmails(4L),
        TotalThreads(5L),
        UnreadThreads(6L),
        myRights = MailboxRights(MayReadItems(true),
          MayAddItems(true),
          MayRemoveItems(true),
          MaySetSeen(true),
          MaySetKeywords(true),
          MayCreateChild(true),
          MayRename(true),
          MayDelete(true),
          MaySubmit(true)),
        IsSubscribed(true),
        MailboxNamespace.personal,
        Rights.EMPTY,
        Quotas(Map())).hasSystemRole must be(false)
    }
    "return true when system" in {
      Mailbox(
        id = TestId.of(42L),
        name = "Inbox",
        parentId = None,
        role = Some(Role.INBOX),
        SortOrder.apply(3L),
        TotalEmails(3L),
        UnreadEmails(4L),
        TotalThreads(5L),
        UnreadThreads(6L),
        MailboxRights(MayReadItems(true),
          MayAddItems(true),
          MayRemoveItems(true),
          MaySetSeen(true),
          MaySetKeywords(true),
          MayCreateChild(true),
          MayRename(true),
          MayDelete(true),
          MaySubmit(true)),
        IsSubscribed(true),
        MailboxNamespace.personal,
        Rights.EMPTY,
        Quotas(Map())).hasSystemRole must be(true)
    }
  }
}
